### DISCUSSION TITLE
Misc function coercing data frames to lists

### DISCUSSION THREAD

**Discussion by ndrubins** [satijalab/seurat#6373] [Q&A] [answered]
Hi,
I'm using `SeuratObject_4.0.4` and `Seurat_4.0.6` and trying to add a `data.frame` (`df` in the example below) as a `misc` to the `Seurat` object:
```
library(Seurat)
df <- data.frame(a = 1:10, b = 10:1, id = paste0("id",1:10))
Seurat::Misc(object = pbmc_small, slot = "testing_df") <- df
```

But when I retrieve `testing_df` it's returned as a `list`:
```
> Seurat::Misc(object = pbmc_small, slot = "testing_df")
$a
 [1]  1  2  3  4  5  6  7  8  9 10

$b
 [1] 10  9  8  7  6  5  4  3  2  1

$id
 [1] "id1"  "id2"  "id3"  "id4"  "id5"  "id6"  "id7"  "id8"  "id9"  "id10"
```

So I'm wondering if there's a way to prevent `Seurat` from coercing the `data.frame` into a `list`?
Another related question is if it's possible to see all the elements under the `misc` `list`?
As it is, I'm getting:
```
> pbmc_small[["RNA"]]@misc
NULL
```

So is it not possible at all to view which elements are saved under the `misc` '`list'?

**Comment by samuel-marsh** [comment, marked as answer]
There appears to be some weirdness with the Misc function from SeuratObject (it should be handling things different but appears not to be.  I'd recommend posting an issue in that repo (https://github.com/mojaveazure/seurat-object/issues).

In meantime though you can run:
```
OBJ@misc[["testing_df"]] <- df
```

The reason that for this 
```
pbmc_small[["RNA"]]@misc
NULL
```
is because unless you specify it to be in the RNA assay the misc data is stored at object level misc slot. So to see what is stored you can run either:

```
Misc(object = OBJ)

#or

OBJ@misc
```

I also have function in my scCustomize package that will take care of storing data in the misc slot.  It automatically checks to make you don't overwrite something with same name already in the misc slot (unless you specify the overwrite param) and everytime you run it will report back the names of each item stored in the misc slot.

```
> OBJ <- Store_Misc_Info_Seurat(seurat_object = OBJ, data_to_store = df, data_name = "testing_df")
Seurat Object now contains the following items in @misc slot:
ℹ 'testing_df'
```

Best,
Sam

**Reply by ndrubins** [reply]
Would you happen to know if it is possible to assign a UMAP `matrix`/`array` to an existing `Seurat' object?
Obviously something like:
```
pbmc_small[["umap"]]@cell.embeddings <- umap.mat
```

Does not work because the the `umap` slot has not been initialized. 

Alternatively, is there a way to initialize a new Seurat object, using the `CreateSeuratObject` function, with an existing UMAP matrix?

**Reply by samuel-marsh** [reply]
Yes you can't just add coordinate that way you need to create a DimReducObject.  See this answer from previous post for details and example code.

https://github.com/satijalab/seurat/issues/4515#issuecomment-846322426

Best,
Sam
